File: packages/android-kotlin/src/main/kotlin/com/cinacoin/chain/TronAdapter.kt
/**
 * TronAdapter — TRON (TRX) chain adapter for Android.
 *
 * Provides TRON-specific operations via TronGrid API:
 * - TRX and TRC-20 balance fetching
 * - Transaction broadcasting
 * - Block retrieval
 * - Energy estimation
 *
 * ## Usage
 * ```kotlin
 * val adapter = TronAdapter("https://api.trongrid.io")
 * val balance = adapter.getBalance("T...")
 * val txId = adapter.sendTransaction("signedTx...")
 * ```
 */
package com.cinacoin.chain

import com.cinacoin.chain.ChainAdapter
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.*
import java.math.BigDecimal
import java.net.HttpURLConnection
import java.net.URL

// ────────────────────────────────────────────────────────────────────────────
// TronGrid API Types
// ────────────────────────────────────────────────────────────────────────────

@Serializable
data class TronAccountResponse(
    val data: List<TronAccount> = emptyList()
)

@Serializable
data class TronAccount(
    val address: String = "",
    val balance: Long = 0,
    val createTime: Long = 0,
    val latestOprationTime: Long = 0,
    val trc20: Map<String, String>? = null,
    val bandwidth: TronBandwidth? = null,
    val assetV2: Map<String, Long>? = null,
    val freeNetLimit: Long = 0,
    val netLimit: Long = 0,
    val energyLimit: Long = 0
)

@Serializable
data class TronBandwidth(
    val freeNetUsed: Long = 0,
    val freeNetLimit: Long = 0,
    val netUsed: Long = 0,
    val netLimit: Long = 0
)

@Serializable
data class TronSendTxResponse(
    val result: Boolean = false,
    val txid: String = "",
    val code: String? = null,
    val message: String? = null
)

@Serializable
data class TronBlockResponse(
    val blockID: String = "",
    val block_header: TronBlockHeader = TronBlockHeader(),
    val transactions: List<TronTxInfo> = emptyList()
)

@Serializable
data class TronBlockHeader(
    val raw_data: TronBlockRawData = TronBlockRawData()
)

@Serializable
data class TronBlockRawData(
    val number: Long = 0,
    val txTrieRoot: String = "",
    val witnessAddress: String = "",
    val parentHash: String = "",
    val version: Int = 0,
    val timestamp: Long = 0
)

@Serializable
data class TronTxInfo(
    val txID: String = "",
    val raw_data: TronTxRawData = TronTxRawData(),
    val signature: List<String> = emptyList()
)

@Serializable
data class TronTxRawData(
    val contract: List<TronContract> = emptyList(),
    val ref_block_bytes: String = "",
    val ref_block_hash: String = "",
    val expiration: Long = 0,
    val timestamp: Long = 0,
    val fee_limit: Long = 0
)

@Serializable
data class TronContract(
    val type: String = "",
    val parameter: TronContractParameter = TronContractParameter()
)

@Serializable
data class TronContractParameter(
    val value: JsonObject = JsonObject(emptyMap()),
    val type_url: String = ""
)

@Serializable
data class TronEnergyResponse(
    val energy_used: Long = 0,
    val energy_limit: Long = 0
)

// ────────────────────────────────────────────────────────────────────────────
// Errors
// ────────────────────────────────────────────────────────────────────────────

sealed class TronError : Exception() {
    data class InvalidAddress(val address: String) : TronError()
    object NotConnected : TronError()
    data class InvalidRpcUrl(val url: String) : TronError()
    data class RpcError(val message: String) : TronError()
    object InvalidParams : TronError()
    data class UnsupportedMethod(val method: String) : TronError()
    data class NotImplemented(val message: String) : TronError()

    override val message: String
        get() = when (this) {
            is InvalidAddress -> "Invalid TRON address: $address"
            NotConnected -> "Not connected to a TRON wallet"
            is InvalidRpcUrl -> "Invalid TronGrid URL: $url"
            is RpcError -> "TronGrid API error: $message"
            InvalidParams -> "Invalid parameters"
            is UnsupportedMethod -> "Unsupported method: $method"
            is NotImplemented -> "Not implemented: $message"
        }
}

// ────────────────────────────────────────────────────────────────────────────
// TRON Adapter
// ────────────────────────────────────────────────────────────────────────────

class TronAdapter(
    private var rpcUrl: String = "https://api.trongrid.io",
    private val apiKey: String? = null,
    private val httpClient: TronHttpClient = DefaultTronHttpClient()
) : ChainAdapter {

    private var _connectedAddress: String? = null

    override val chainName: String = "tron"

    override var endpoint: String
        get() = rpcUrl
        set(value) { rpcUrl = value }

    override val isConnected: Boolean
        get() = _connectedAddress != null

    override val connectedAddress: String?
        get() = _connectedAddress

    companion object {
        val MAINNET = "https://api.trongrid.io"
        val NILE = "https://nile.trongrid.io"
        val SHASTA = "https://api.shasta.trongrid.io"

        /** Validate a TRON base58check address. */
        fun isValidAddress(address: String): Boolean {
            return address.startsWith("T") && address.length == 34 &&
                address.all { it.isLetterOrDigit() }
        }

        /** Convert sun to TRX string. */
        fun sunToTrx(sun: Long): String =
            BigDecimal(sun).divide(BigDecimal("1000000"), 6, java.math.RoundingMode.HALF_UP)
                .toPlainString()
    }

    private val headers: Map<String, String>
        get() = if (apiKey != null) mapOf("TRON-PRO-API-KEY" to apiKey) else emptyMap()

    // ─── Balance ────────────────────────────────────────────────────────

    /**
     * Get TRX balance for an address via TronGrid API.
     * @return Balance in TRX as decimal string.
     */
    override suspend fun getBalance(address: String): String =
        withContext(Dispatchers.IO) {
            require(isValidAddress(address)) { "Invalid TRON address: $address" }

            val url = "$rpcUrl/v1/accounts/$address"
            val responseBody = httpClient.get(url, headers)
            val response = Json.decodeFromString<TronAccountResponse>(responseBody)

            if (response.data.isEmpty()) {
                return@withContext "0.000000"
            }

            val account = response.data.first()
            sunToTrx(account.balance)
        }

    /**
     * Get TRC-20 token balance for an address.
     * @param tokenContract TRC-20 contract address.
     */
    suspend fun getTokenBalance(address: String, tokenContract: String): String =
        withContext(Dispatchers.IO) {
            val url = "$rpcUrl/v1/accounts/$address"
            val responseBody = httpClient.get(url, headers)
            val response = Json.decodeFromString<TronAccountResponse>(responseBody)

            if (response.data.isEmpty()) return@withContext "0"

            response.data.first().trc20?.get(tokenContract) ?: "0"
        }

    // ─── Send Transaction ───────────────────────────────────────────────

    /**
     * Send a signed transaction via TronGrid API.
     * @param signedTx JSON-encoded signed transaction.
     * @return Transaction ID.
     */
    override suspend fun sendTransaction(signedTx: String): String =
        withContext(Dispatchers.IO) {
            val responseBody = httpClient.post("$rpcUrl/wallet/broadcasttransaction", signedTx, headers)
            val response = Json.decodeFromString<TronSendTxResponse>(responseBody)

            if (!response.result) {
                throw TronError.RpcError(response.message ?: "Transaction broadcast failed")
            }

            response.txid
        }

    // ─── Block Number ──────────────────────────────────────────────────

    /**
     * Get the latest block number.
     */
    override suspend fun getLatestBlock(): Long = withContext(Dispatchers.IO) {
        val url = "$rpcUrl/blocks?limit=1"
        val responseBody = httpClient.get(url, headers)
        val obj = Json.parseToJsonElement(responseBody).jsonArray

        if (obj.isEmpty()) throw TronError.RpcError("No blocks found")

        obj[0].jsonObject["block_header"]?.jsonObject?.get("raw_data")
            ?.jsonObject?.get("number")?.jsonPrimitive?.longOrNull
            ?: throw TronError.RpcError("Invalid block response")
    }

    /**
     * Get block by number.
     */
    suspend fun getBlockByNumber(number: Long): TronBlockResponse =
        withContext(Dispatchers.IO) {
            val url = "$rpcUrl/blocks?limit=1&min=$number&max=$number"
            val responseBody = httpClient.get(url, headers)
            val arr = Json.parseToJsonElement(responseBody).jsonArray
            if (arr.isEmpty()) throw TronError.RpcError("Block $number not found")
            Json.decodeFromString<TronBlockResponse>(arr[0].toString())
        }

    // ─── Fee / Energy Estimation ───────────────────────────────────────

    /**
     * Estimate energy cost for a transaction.
     */
    override suspend fun estimateFee(): String = withContext(Dispatchers.IO) {
        // TRON bandwidth: 1 transaction uses ~268 bandwidth
        // Energy varies by contract interaction
        "268" // Bandwidth cost for basic TRX transfer
    }

    /**
     * Estimate energy for a contract call.
     * @param ownerAddress Sender address.
     * @param contractAddress Contract address.
     * @param functionSelector Function selector (e.g. "transfer(address,uint256)").
     */
    suspend fun estimateEnergy(
        ownerAddress: String,
        contractAddress: String,
        functionSelector: String = "transfer(address,uint256)",
        parameter: String = ""
    ): String = withContext(Dispatchers.IO) {
        val body = buildJsonObject {
            put("owner_address", JsonPrimitive(ownerAddress))
            put("contract_address", JsonPrimitive(contractAddress))
            put("function_selector", JsonPrimitive(functionSelector))
            if (parameter.isNotEmpty()) put("parameter", JsonPrimitive(parameter))
        }
        val requestBody = Json.encodeToString(body)
        val responseBody = httpClient.post("$rpcUrl/wallet/estimateenergy", requestBody, headers)
        val obj = Json.parseToJsonElement(responseBody).jsonObject
        obj["energy_required"]?.jsonPrimitive?.longOrNull?.toString() ?: "0"
    }

    // ─── Account Resources ──────────────────────────────────────────────

    /**
     * Get account bandwidth and energy resources.
     */
    suspend fun getAccountResources(address: String): JsonElement =
        withContext(Dispatchers.IO) {
            val body = """{"address":"$address"}"""
            val responseBody = httpClient.post("$rpcUrl/wallet/getaccountresource", body, headers)
            Json.parseToJsonElement(responseBody)
        }

    /**
     * Get transaction info by ID.
     */
    suspend fun getTransaction(txId: String): JsonElement =
        withContext(Dispatchers.IO) {
            val url = "$rpcUrl/v1/transactions/$txId"
            val responseBody = httpClient.get(url, headers)
            Json.parseToJsonElement(responseBody)
        }

    // ─── Connection Management ──────────────────────────────────────────

    override fun setConnectedAddress(address: String) {
        require(isValidAddress(address)) { "Invalid TRON address: $address" }
        _connectedAddress = address
    }

    override fun disconnect() {
        _connectedAddress = null
    }

    // ─── Unified Request ────────────────────────────────────────────────

    suspend fun request(method: String, params: List<Any?> = emptyList()): JsonElement =
        when (method) {
            "tron_getBalance" -> JsonPrimitive(getBalance(params.firstOrNull() as? String ?: ""))
            "tron_sendTransaction" -> JsonPrimitive(sendTransaction(params.firstOrNull() as? String ?: ""))
            "tron_getBlockNumber" -> JsonPrimitive(getLatestBlock())
            "tron_estimateFee" -> JsonPrimitive(estimateFee())
            else -> throw TronError.UnsupportedMethod(method)
        }
}

// ────────────────────────────────────────────────────────────────────────────
// HTTP Client Interface
// ────────────────────────────────────────────────────────────────────────────

interface TronHttpClient {
    suspend fun get(url: String, headers: Map<String, String>): String
    suspend fun post(url: String, body: String, headers: Map<String, String>): String
}

class DefaultTronHttpClient : TronHttpClient {
    override suspend fun get(url: String, headers: Map<String, String>): String =
        withContext(Dispatchers.IO) {
            val connection = URL(url).openConnection() as HttpURLConnection
            try {
                connection.requestMethod = "GET"
                headers.forEach { connection.setRequestProperty(it.key, it.value) }
                connection.connectTimeout = 15_000
                connection.readTimeout = 30_000
                if (connection.responseCode != HttpURLConnection.HTTP_OK) {
                    throw TronError.RpcError("HTTP ${connection.responseCode}")
                }
                connection.inputStream.bufferedReader().use { it.readText() }
            } finally {
                connection.disconnect()
            }
        }

    override suspend fun post(url: String, body: String, headers: Map<String, String>): String =
        withContext(Dispatchers.IO) {
            val connection = URL(url).openConnection() as HttpURLConnection
            try {
                connection.requestMethod = "POST"
                connection.setRequestProperty("Content-Type", "application/json")
                headers.forEach { connection.setRequestProperty(it.key, it.value) }
                connection.doOutput = true
                connection.connectTimeout = 15_000
                connection.readTimeout = 30_000
                connection.outputStream.write(body.toByteArray(Charsets.UTF_8))
                connection.outputStream.flush()
                if (connection.responseCode != HttpURLConnection.HTTP_OK) {
                    throw TronError.RpcError("HTTP ${connection.responseCode}")
                }
                connection.inputStream.bufferedReader().use { it.readText() }
            } finally {
                connection.disconnect()
            }
        }
}

File: onux/packages/unity-csharp/Runtime/Chain/TronAdapter.cs
using System;
using System.Numerics;
using System.Text;
using System.Threading.Tasks;
using System.Collections.Generic;
using UnityEngine.Networking;
using Newtonsoft.Json;

namespace Cinacoin.Chain
{
    /// <summary>
    /// Tron chain adapter.
    /// Uses TronGrid API for balance, transaction, and block queries.
    /// Compatible with Tron mainnet and testnet (Shasta).
    /// </summary>
    public class TronAdapter : IChainAdapter
    {
        private string _rpcUrl;
        private string _apiKey;
        private static int _rpcId;

        public string Namespace => "tron";
        public string ChainReference => "mainnet";
        public string ChainId => "tron:mainnet";
        public string ChainName => "Tron Mainnet";

        public TronAdapter(string apiKey = null) { _apiKey = apiKey; }

        public void Init(string rpcUrl)
        {
            _rpcUrl = string.IsNullOrEmpty(rpcUrl) ? "https://api.trongrid.io" : rpcUrl;
        }

        public async Task<BigInteger> GetBalanceAsync(string address)
        {
            var url = $"{_rpcUrl}/v1/accounts/{NormalizeAddress(address)}";
            var response = await CallJsonAsync<TronAccountResponse>(url);
            var balance = response.Data[0].Balance;
            return new BigInteger(balance);
        }

        public async Task<string> GetBalanceFormattedAsync(string address)
        {
            var sun = await GetBalanceAsync(address);
            return ((double)sun / 1e6).ToString("F6");
        }

        public async Task<long> GetLatestBlockAsync()
        {
            var url = $"{_rpcUrl}/wallet/block/now";
            var response = await PostJsonAsync<TronBlock>(url, new { });
            return response.BlockHeader.RawData.Number;
        }

        public async Task<BigInteger> EstimateFeeAsync(string from, string to, string data = null)
        {
            // Tron bandwidth/energy fee estimation
            // Base fee for a TRX transfer is ~0 TRX (bandwidth only)
            // Smart contract calls consume energy
            if (string.IsNullOrEmpty(data))
                return new BigInteger(0); // Free bandwidth for TRX transfers

            // Estimate energy for contract call (rough estimate)
            return new BigInteger(65000); // ~65,000 energy units
        }

        public async Task<string> SendTransactionAsync(string signedTx)
        {
            var url = $"{_rpcUrl}/wallet/broadcasttransaction";
            var response = await PostJsonAsync<TronBroadcastResult>(url, signedTx);
            return response.TxId;
        }

        public async Task<string> GetTransactionAsync(string txHash)
        {
            var url = $"{_rpcUrl}/wallet/gettransactionbyid";
            var body = new { value = txHash };
            return await PostJsonAsync<string>(url, body);
        }

        /// Broadcast a signed transaction.
        public async Task<TronBroadcastResult> BroadcastTransactionAsync(TronTransaction signedTx)
        {
            var url = $"{_rpcUrl}/wallet/broadcasttransaction";
            return await PostJsonAsync<TronBroadcastResult>(url, signedTx);
        }

        public bool IsValidAddress(string address)
        {
            if (string.IsNullOrEmpty(address)) return false;
            // Tron addresses start with 'T' and are 34 characters (base58)
            if (address.StartsWith("T") && address.Length == 34) return true;
            // Hex format: 42 characters starting with 0x41
            if (address.StartsWith("41") && address.Length == 42) return true;
            return false;
        }

        private string NormalizeAddress(string address)
        {
            // Convert hex address to base58 if needed
            if (address.StartsWith("0x")) address = address.Substring(2);
            return address;
        }

        private async Task<T> CallJsonAsync<T>(string url)
        {
            using var request = new UnityWebRequest(url, "GET")
            {
                downloadHandler = new DownloadHandlerBuffer(),
                timeout = 30
            };
            if (!string.IsNullOrEmpty(_apiKey))
                request.SetRequestHeader("TRON-PRO-API-KEY", _apiKey);

            var operation = request.SendWebRequest();
            while (!operation.isDone) await Task.Delay(1);

            if (request.result != UnityWebRequest.Result.Success)
                throw new Exception($"Tron request failed: {request.error}");

            return JsonConvert.DeserializeObject<T>(request.downloadHandler.text);
        }

        private async Task<T> PostJsonAsync<T>(string url, object body)
        {
            var json = JsonConvert.SerializeObject(body);
            using var request = new UnityWebRequest(url, "POST")
            {
                uploadHandler = new UploadHandlerRaw(Encoding.UTF8.GetBytes(json)),
                downloadHandler = new DownloadHandlerBuffer(),
                timeout = 30
            };
            request.SetRequestHeader("Content-Type", "application/json");
            if (!string.IsNullOrEmpty(_apiKey))
                request.SetRequestHeader("TRON-PRO-API-KEY", _apiKey);

            var operation = request.SendWebRequest();
            while (!operation.isDone) await Task.Delay(1);

            if (request.result != UnityWebRequest.Result.Success)
                throw new Exception($"Tron POST failed: {request.error}");

            return JsonConvert.DeserializeObject<T>(request.downloadHandler.text);
        }
    }

    // ─── Tron API Response Types ────────────────────────────────────────

    internal class TronAccountResponse
    {
        [JsonProperty("data")]
        public List<TronAccount> Data { get; set; }
    }

    internal class TronAccount
    {
        [JsonProperty("balance")]
        public long Balance { get; set; }
        [JsonProperty("address")]
        public string Address { get; set; }
    }

    internal class TronBlock
    {
        [JsonProperty("block_header")]
        public TronBlockHeader BlockHeader { get; set; }
    }

    internal class TronBlockHeader
    {
        [JsonProperty("raw_data")]
        public TronBlockRawData RawData { get; set; }
    }

    internal class TronBlockRawData
    {
        [JsonProperty("number")]
        public long Number { get; set; }
    }

    internal class TronBroadcastResult
    {
        [JsonProperty("txid")]
        public string TxId { get; set; }
        [JsonProperty("result")]
        public bool Result { get; set; }
        [JsonProperty("message")]
        public string Message { get; set; }
    }

    internal class TronTransaction
    {
        [JsonProperty("visible")]
        public bool Visible { get; set; }
        [JsonProperty("txID")]
        public string TxId { get; set; }
        [JsonProperty("raw_data")]
        public Dictionary<string, object> RawData { get; set; }
        [JsonProperty("raw_data_hex")]
        public string RawDataHex { get; set; }
        [JsonProperty("signature")]
        public List<string> Signature { get; set; }
    }
}
